support for esp32 with freertos

donRaphaco opened the pull request.

ralph-lange commented:
Relates to https://github.com/micro-ROS/freertos_apps/pull/12

pablogs9 commented:
Hello @ralph-lange and @donRaphaco what do you think about adding this port to Foxy instead of Dashing?

ralph-lange commented:
I think even both should be doable. Let's try it, @donRaphaco.

pablogs9 commented:
I'm going to test this tomorrow. We have here [this model](https://www.mouser.es/ProductDetail/Espressif-Systems/ESP32-DevKitC-32D?qs=%252BEew9%252B0nqrDsObWEpDx6YQ==&vip=1&gclid=CjwKCAjwx9_4BRAHEiwApAt0zp-wKcTROecqGSiUayTDMll12aH3rnxjGcOj_FHgJwg9yRyWokB52BoCuPkQAvD_BwE). Do you think that is going to work @donRaphaco? 

In any case if my try is successful, I'll check the migration to foxy.

donRaphaco commented:
> 
> 
> I'm going to test this tomorrow. We have here [this model](https://www.mouser.es/ProductDetail/Espressif-Systems/ESP32-DevKitC-32D?qs=%252BEew9%252B0nqrDsObWEpDx6YQ==&vip=1&gclid=CjwKCAjwx9_4BRAHEiwApAt0zp-wKcTROecqGSiUayTDMll12aH3rnxjGcOj_FHgJwg9yRyWokB52BoCuPkQAvD_BwE). Do you think that is going to work @donRaphaco?
> 
> In any case if my try is successful, I'll check the migration to foxy.

Yes, I think this board should work.

pablogs9 commented:
Hello @donRaphaco, I'm starting the test. In the creation step it says:

```
Installing ESP-IDF tools
/usr/bin/env: ‘python’: No such file or directory
```

is the python2.7 dependency compulsory? In micro-ROS/ROS Foxy the Ubuntu version is 20.04 and it only has python3 since the 2.7 version is deprecated. Is there any simple workaround?

pablogs9 commented:
What about adding `firmware/freertos_apps/microros_esp32_extensions/build` and `firmware/freertos_apps/microros_esp32_extensions/sdkconfig` to `firmware/freertos_apps/microros_esp32_extensions/.gitignore`?

pablogs9 commented:
I'm having troubles with some compiler flags during the colcon build step inside the the `builld_firmware.sh` step.

I would like to know if our GCC version mismatch. Here my `configure_firmware.sh` output:

```
vscode@pgarrido:/workspaces/foxy_ws/uros_freertos_esp32$ ros2 run micro_ros_setup configure_firmware.sh int32_publisher -t serial -d 0
Configuring firmware for freertos platform esp32
Using serial device USART0.
Configured serial mode with agent at USART0
-- Found Git: /usr/bin/git (found version "2.25.1") 
-- IDF_TARGET not set, using default target: esp32
-- The C compiler identification is GNU 8.2.0
-- The CXX compiler identification is GNU 8.2.0
```

Is your compiler also ` GNU 8.2.0`?

pablogs9 commented:
Also pyserial required for 20.04:

`pip install pyserial`

donRaphaco commented:
Hi @pablogs9,
I installed Ubuntu 20.04 myself yesterday and I'm currently trying it on ros2 foxy.
With the fresh installation I encountered the same problems as you described.

> is the python2.7 dependency compulsory? In micro-ROS/ROS Foxy the Ubuntu version is 20.04 and it only has python3 since the 2.7 version is deprecated. Is there any simple workaround?

The esp-idf toolchain actually supports python2 and 3 but the install scripts use the `#! /usr/bin/env python` shebang, which defaults to python2 on ubuntu.
I'm working to fix it, I'll come back to you when everything works on foxy.

pablogs9 commented:
Nice, thank you so much @donRaphaco

donRaphaco commented:
@pablogs9  I updated the PR to foxy and fixed the "python not found" error.
There were 2 dependencies which I needed to install with apt: `gperf` and `python3-venv`.

One more thing as I noticed you want to use serial port 0:
Note that the esp uses serial port 0 for console output by default, if you want to use this port for microros you can change this setting in `menuconfig` which is accessible  by running `ros2 run micro_ros_setup build_firmware.sh menuconfig` then go to `Component config` -> `Common ESP-related` -> `UART for console output`.

donRaphaco commented on micro_ros_setup/config/freertos/esp32/create.sh:
```diff
@@ -0,0 +1,38 @@
+    # Install toolchain
+    mkdir toolchain
+
+    # install prerequesites for esp-idf
+    # sudo apt-get install -y gperf python3-venv
```
I'm not sure if I should run this `sudo` command here to install apt dependencies.
Or is it better to just check if the dependencies are satisfied and if not request the user to install them themself?

pablogs9 replied:
are they available using rosdep?

donRaphaco replied:
yes they are should I put them into package.xml?

pablogs9 replied:
yes, they should be in the `package.xml` inside `micro-ros-build/micro_ros_setup/config/freertos/esp32/`. This way, the build system should resolve this dependecies.

Also, can you add this platform to the CI.yml?

donRaphaco replied:
Ok I will.

donRaphaco replied:
Done :)
The CI step for esp32 still fails because micro-ROS/freertos_apps/pull/12 is not merged to foxy.

pablogs9 commented:
Hello @donRaphaco is it required to have `CONFIG_ESP_MAIN_TASK_STACK_SIZE=25000` and then create `xTaskCreate(appMain, "uros_task", 12*2048, NULL, 5, NULL);`

Is possible that at least `CONFIG_ESP_MAIN_TASK_STACK_SIZE` can be configured with less stack?

pablogs9 commented:
Hi, again @donRaphaco I have achieved run this with UDP transport over WiFi. I had (and still have) some troubles with includes in the CMake toolchain. I have fixed this issue by commenting, but I think that we can do an elegant solution in order to have ESP32 Wifi and Serial transports working.

Tomorrow I'll commit these changes to a new branch in `freertos_apps` and to this repo, and I'll ask you if you can help me with this include mess.

pablogs9 commented:
Hello @donRaphaco @ralph-lange can you give a try to the WiFi mode?

In order to try it, you have to checkout this branch and this other branch after creation step is done: https://github.com/micro-ROS/freertos_apps/tree/feature/esp32_wifi

Once the app is configured, by using the `ros2 run micro_ros_setup build_firmware.sh menuconfig` and under micro-ROS setting, you can configure your WiFI SSID and PSK.

For us the result is really successful, this morning I have seen topic rates (single Int32) of 300 Hz using UDP.

I still have some issues with LwIP includes, [this patch](https://github.com/micro-ROS/freertos_apps/blob/feature/esp32_wifi/microros_esp32_extensions/lwip.patch) is my workaround. Please tell me if you know some solutions for this include mess.

Thanks!!

donRaphaco commented:
Hi @pablogs9, great that you got UDP to work!
For the rest of the week I'm quite busy, but I hope next week I can have a look and try it.
Also I'll find a better value for `CONFIG_ESP_MAIN_TASK_STACK_SIZE` which is quite high, because at first I ran the app in the main task and forgot about this setting.

pablogs9 commented:
Nice, let me know your impressions next week and we will merge these PRs.

Regarding the `CONFIG_ESP_MAIN_TASK_STACK_SIZE`,  I'm going to change it in my new branch, don't worry about that.

donRaphaco commented:
Hi @pablogs9 I tested UDP and WiFi with your changes and everything works 👍.
Also I found out that the patch is not needed if you add this line to the toolchain.cmake:
`add_definitions(-DLWIP_IPV4 -DLWIP_IPV6)`
